### PR TITLE
Hw 24

### PR DESCRIPTION
**Pull request by ilya449**

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -25,6 +25,11 @@
...
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>org.projectlombok</groupId>
+            <artifactId>lombok</artifactId>
+            <version>1.18.10</version>
+        </dependency>
     </dependencies>
     <build>
         <plugins>
```
Comment (pavlokhomiak) [inline pom.xml:35]: Add checkstyle plugin into your pom.xml to keep your code clean

File: src/main/java/com/internet/shop/dao/impl/ProductDaoImpl.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    }
+
+    @Override
+    public Product update(Product product) {
+        return Storage.getProducts().stream()
+                .filter(p -> p.getId().equals(product.getId()))
+                .map(p -> p = product)
+                .findFirst()
+                .orElseThrow(() ->
+                        new NoSuchElementException(String.format(EXCEPTION_MESSAGE,
```
Comment (pavlokhomiak) [inline src/main/java/com/internet/shop/dao/impl/ProductDaoImpl.java:41]: no exceptions. return product

File: src/main/java/com/internet/shop/dao/impl/ProductDaoImpl.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+                                product.getId().toString())));
+    }
+
+    @Override
+    public boolean delete(Long id) {
+        return Storage.getProducts().remove(get(id).orElseThrow(() ->
+                new NoSuchElementException(String.format(EXCEPTION_MESSAGE, id.toString()))));
```
Comment (pavlokhomiak) [inline src/main/java/com/internet/shop/dao/impl/ProductDaoImpl.java:48]: return false

File: pom.xml (hunk 1 above)
Comment (ilya449) [reply pom.xml:35]: I have it in master so here it isn't highlighted. If you expand my pom.xml file you can make sure

File: src/main/java/com/internet/shop/dao/impl/ProductDaoImpl.java [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    public Product update(Product product) {
+        Storage.getProducts().stream()
+                .filter(p -> p.getId().equals(product.getId()))
+                .forEach(p -> p = product);
+        return product;
+    }
+
+    @Override
+    public boolean delete(Long id) {
+        return Storage.getProducts().removeIf(p -> p.getId().equals(id));
```
Comment (dmytro-hryhoruk) [inline src/main/java/com/internet/shop/dao/impl/ProductDaoImpl.java:40]:
```suggestion
        return Storage.getProducts()
        .removeIf(p -> p.getId().equals(id));
```

**Review by dmytro-hryhoruk — APPROVED** [review]

File: src/main/java/com/internet/shop/dao/impl/ProductDaoImpl.java [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+    public List<Product> getAll() {
+        return Storage.getProducts();
+    }
+
+    @Override
+    public Product update(Product product) {
+        Storage.getProducts().stream()
+                .filter(p -> p.getId().equals(product.getId()))
+                .forEach(p -> p = product);
+        return product;
```
Comment (dmytro-hryhoruk) [inline src/main/java/com/internet/shop/dao/impl/ProductDaoImpl.java:35]: does it work?

Comment (ilya449) [reply src/main/java/com/internet/shop/dao/impl/ProductDaoImpl.java:35]: yes

Comment (ilya449) [reply src/main/java/com/internet/shop/dao/impl/ProductDaoImpl.java:35]: Maybe this variant will be better?
```
.forEach(p -> {
                    p.setName(product.getName());
                    p.setPrice(product.getPrice());
                });
```
But both of them work

Comment (ilya449) [reply src/main/java/com/internet/shop/dao/impl/ProductDaoImpl.java]: No, I made a mistake when tested and missed it. But I've fixed it and now it works well.
Is it a good practice to create some method like ```updateFields(Product product)``` in class "Product" and invoke it here: ```forEach(p -> p.updateFields(product))```?

File: src/main/java/com/internet/shop/dao/impl/ProductDaoImpl.java [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+    public List<Product> getAll() {
+        return Storage.getProducts();
+    }
+
+    @Override
+    public Product update(Product product) {
+        Storage.getProducts().stream()
+                .filter(p -> p.getId().equals(product.getId()))
+                .forEach(p -> Storage.getProducts()
+                        .set(Storage.getProducts().indexOf(p), product));
```
Comment (dmytro-hryhoruk) [inline src/main/java/com/internet/shop/dao/impl/ProductDaoImpl.java:35]: let's use IntStream and you will have indexes to access list elemets

Comment (ilya449) [reply src/main/java/com/internet/shop/dao/impl/ProductDaoImpl.java]: Ok. But could you explain why this variant is better than my previous?

Comment (dmytro-hryhoruk) [reply src/main/java/com/internet/shop/dao/impl/ProductDaoImpl.java]: indexOf() will once again iterate through the whole list to find the index of this element. Using integer stream you will have indexes right away without additional iteration.

**Review by dmytro-hryhoruk — APPROVED** [review]

**Pull request merged**